File: contents/05/basic-2/solution.py
def partition(A, p, r):
    x = A[r]
    i = p-1
    for j in range(p, r):
        if A[j] <= x:
            i += 1
            A[i],A[j] = A[j], A[i]
    A[i+1], A[r] = A[r], A[i+1]
    return i+1 



def quicksort(A, p, r):
    if p < r:
        q = partition(A,p,r)
        quicksort(A,p, q-1)
        quicksort(A,q+1, r)

N = int(input())
A = list(map(int, input().split()))

quicksort(A, 0, N-1)
A = [str(a) for a in A]
print(" ".join(A))

File: contents/05/basic-2/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i, m, n) for (int i = (m); i < (n); i++)
int partition(int A[], int p, int r)
{
	int x = A[r];
	int i = p - 1;
	rep(j, p, r)
	{
		if (A[j] <= x)
		{
			i++;
			swap(A[i], A[j]);
		}
	}
	swap(A[i + 1], A[r]);
	return i + 1;
}

void quicksort(int A[], int p, int r)
{
	if (p < r)
	{
		int q = partition(A, p, r);
		quicksort(A, p, q - 1);
		quicksort(A, q + 1, r);
	}
}

int main()
{
	int N;
	int A[100010];
	cin >> N;
	rep(i, 0, N)
	{
		cin >> A[i];
	}

	quicksort(A, 0, N - 1);
	rep(i, 0, N)
	{
		cout << A[i] << " \n"[i == N - 1];
	}
}